crates/spartan-farmer/src/commands/dht/client.rs

```diff
@@ -0,0 +1,308 @@
+use super::*;
+use super::{core::create_node, eventloop::EventLoop};
+use libp2p::kad::QueryInfo;
+
+#[derive(Copy, Clone, Debug)]
+pub enum ClientType {
+    // Bootstrap node. It uses the following fields from `ClientConfig`:
+    // 1. `bootstrap_keys`: Private keys/private key location to create bootstrap node peerId.
+    // 2. `listen_addr`: Listening address for Bootstrap node.
+    Bootstrap,
+    // Normal node. It uses the following fields from `ClientConfig`:
+    // 1. `bootstrap_nodes`: Bootstrap nodes addresses that the normal node must connect to.
+    // For setting listening address, use client.start_listening.
+    Normal,
+}
+
+pub enum ClientEvent {
+    // Event for adding a listening address.
+    Listen {
+        addr: Multiaddr,
+        sender: oneshot::Sender<OneshotType>,
+    },
+    // Kademlia Random Walk for Peer Discovery. (GetClosestPeer)
+    #[allow(dead_code)]
+    RandomWalk {
+        key: Option<PeerId>,
+        sender: oneshot::Sender<QueryId>,
+    },
+    // List all known peers.
+    #[allow(dead_code)]
+    KnownPeers {
+        sender: oneshot::Sender<Vec<PeerId>>,
+    },
+    // Dial another peer.
+    #[allow(dead_code)]
+    Dial {
+        addr: Multiaddr,
+        peer: PeerId,
+        sender: oneshot::Sender<OneshotType>,
+    },
+    // Bootstrap during the initial connection to the DHT.
+    // NOTE: All the bootstrap nodes must already be connected to the swarm before we can start the
+    // bootstrap process.
+    Bootstrap {
+        sender: oneshot::Sender<QueryId>,
+    },
+    // Get all listening addresses.
+    #[allow(dead_code)]
+    Listeners {
+        sender: oneshot::Sender<Vec<Multiaddr>>,
+    },
+    // Read Kademlia Query Result.
+    #[allow(dead_code)]
+    QueryResult {
+        qid: QueryId,
+        sender: oneshot::Sender<String>,
+    },
+}
+
+pub struct ClientConfig {
+    pub bootstrap_nodes: Vec<String>, // Vec<(Multiaddr, PeerId)>,
+    pub client_type: ClientType,
+    pub listen_addr: Option<Multiaddr>,
+}
+
+pub struct Client {
+    pub peerid: PeerId,
+    // This channel sends events from Client to EventLoop.
+    client_tx: Sender<ClientEvent>,
+}
+
+impl Client {
+    pub fn new(peerid: PeerId, client_tx: Sender<ClientEvent>) -> Self {
+        Client { peerid, client_tx }
+    }
+
+    // Read the Query Result for a specific Kademlia query.
+    // This method returns information about pending as well as finsihed queries.
+    #[allow(dead_code)]
```

Comment:
The number of `#[allow(dead_code)]` is worrying, please remove all that are not necessary and add TODOs about those that should stay explaining why that is the case.